enhancements/cluster-scope-secret-volumes/shared-resource-validation.md

```diff
@@ -0,0 +1,413 @@
+#### Failure Modes
+
+1. Cluster Storage Operator / Shared Resource CSI Driver Operator failures could result in the ValidatingAdmissionWebhook not getting started
+2. API Server failures could result in the ValidatingAdmissionWebhook not getting called
+3. Unanticipated bugs in the ValidatingAdmissionWebhook result in incorrect validations
+
+Each of those ValidatingAdmissionWebhook checks either already are or will be made in the Shared Resource Controller and CSI Driver.
+So no violations of what we allow should occur, unless there are bugs across the entire path from webhook to controller to CSI driver for 
+a given constraint.
+
+Incorrectly preventing the create/update of Shared Resources at the moment appears to be relatively minor on the problem
+scale for cluster administrators and support.  But the error returned wil be directly available to the user.  That, along
+with the CLI invocation used, and any associated YAML for the object, should lead to straight forward diagnostics for 
+support.
+
+The enforcement of 'readOnly: true' in particular is worth discussing in this section.  It seems like very simple 
+pattern matching / field checking, but unanticipated bugs when performing this check
+in a ValidatingAdmissionWebhook could be impactful, as we are ultimately gating whether a Pod is allowed or not.  If 
+there is a bug which results in valid Pods being rejected, that could have broad and major consequences.
+
+Does that risk give enough pause so that we punt on the user story around more useful failure modes when the user has 
+failed to set 'readOnly: true' on their Pod's CSI Volume mount?  The fact that there is general agreement on pursuing 
+this EP, the assumption is the answer to that question is "no, we'll just be careful to get it right".  But let's 
+discuss as needed.
+
+The OCP teams possibly involved with issues:
+- API Server (webhook invocation, confusion around creating/modifying objects)
+- Kubelet/node (incorrectly perhaps if valid Pods fail to come up, unless there is a problem beyond this EP's validations)
+- Cluster Storage Operator (currently storage team, if a CSO problem leads to the webhook not coming up) 
+- Shared Resource Operator/Driver themselves (currently build/jenkins team)
+
+
+#### Support Procedures
+
+##### Detect Failure modes and Possible Symptoms
+
+- If the webhook is not running, kube-apiserver logs will show errors like "failed to call admission webhook xyz".
+- Cluster Storage Operator / CSI Shared Resource Operator will degrade with message "Failed to launch webhook server" and reason "WehhookServerFailed".
+- The k8s metric `webhook_admission_duration_seconds("<name of our new admission controller>", "validating", "put", "false")`
+    will show >1s latency and alert `WebhookAdmissionLatencyHigh` will fire.
+- Otherwise, we should be able to fall back on must gather.
+
+##### Disable the API extension (e.g. remove ValidatingWebhookConfiguration `<names of the our new webhook configs>`)
```

Comment:
Is this achievable - won't our operator deploy the webhook and reconcile the `ValidatingWebhookConfiguration` object?

Reply:
Well I'm actually just following the template, trying to address the questions it asks.  I just to converted a bullet point in the template to a heading for readability.

Can't seem to do line numbers when referencing lines in markdown files, but if you search for `disable the API extension (e.g. remove MutatingWebhookConfiguration xyz, remove APIService foo)` in https://github.com/openshift/enhancements/blob/db2ce78a020c2998096f7763bddbe108e40ce3ca/guidelines/enhancement_template.md you'll see the question.

As to "achievability", other than disabling the CSO (I imagine that has the same consequence support wise as disabling the CVO, but don't know that with certainty) is the only means for that I can think of.

Maybe for crazy critical support situations we will suggest that to customers to prevent clusters from going belly up ??

Reply:
This might be the best option we have in terms of "emergency support":
- Scale the CVO deployment to 0
- Scale the cluster-storage-operator deployment to 0
- Remove the ValidatingAdmissionWebhook configuration.